recipes/dbus/1.x.x/conanfile.py

```diff
@@ -0,0 +1,110 @@
+        "enable_checks": [True, False],
+        "install_system_libs": [True, False]}
+
+    default_options = {
+        "fPIC": True,
+        "with_x11": False,
+        "with_glib": False,
+        "enable_assert": True,
+        "enable_checks": True,
+        "install_system_libs": False}
+
+    generators = "cmake", "cmake_find_package", "cmake_paths"
+
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+    _cmake = None
+
+    def configure(self):
+        if self.settings.os == 'Windows':
+            raise ConanInvalidConfiguration("D-Bus is not compatible with Windows")
+        if self.settings.os == "Macos":
+            raise ConanInvalidConfiguration("D-Bus is not compatible with MacOS")
+
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def requirements(self):
+        self.requires("expat/2.2.10")
+        if self.options.with_glib:
+            self.requires("glib/2.67.0")
+
+        if self.options.with_x11:
+            self.requires("xorg/system")
+
+    def _configure_cmake(self):
+        if not self._cmake:
+            self._cmake = CMake(self)
+
+            self._cmake.definitions["DBUS_BUILD_TESTS"] = False
+            self._cmake.definitions["DBUS_ENABLE_DOXYGEN_DOCS"] = False
+            self._cmake.definitions["DBUS_ENABLE_XML_DOCS"] = False
+
+            self._cmake.definitions["DBUS_BUILD_X11"] = self.options.with_x11
+            self._cmake.definitions["DBUS_WITH_GLIB"] = self.options.with_glib
+            self._cmake.definitions["DBUS_DISABLE_ASSERT"] = not self.options.enable_assert
+            self._cmake.definitions["DBUS_DISABLE_CHECKS"] = not self.options.enable_checks
+            self._cmake.definitions["DBUS_INSTALL_SYSTEM_LIBS"] = self.options.install_system_libs
+
+            path_to_cmake_lists = os.path.join(self._source_subfolder, "cmake")
+
+            self._cmake.configure(source_folder=path_to_cmake_lists,
+                                  build_folder=self._build_subfolder)
+        return self._cmake
+
+    def build(self):
+        dbus_cmake = tools.os.path.join(
+            self._source_subfolder, "cmake", "CMakeLists.txt")
+
+        if self.options.with_glib:
+            tools.replace_in_file(dbus_cmake, "GLib2", "glib")
+            tools.replace_in_file(dbus_cmake, "GLIB2", "GLIB")
+
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy(pattern="COPYING", dst="licenses",
+                  src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        for i in ["var", "share", "etc"]:
+            shutil.move(os.path.join(self.package_folder, i), os.path.join(self.package_folder, "res", i))
```

Comment:
Instead of moving, you can set `CMAKE_INSTALL_DATAROOTDIR` to  `os.path.join(self.package_folder, "res", "shared"` in build.
See cmake's GNUInstallDirs for more variables.

Also, are the files in var, share and etc needed for development/running?
Can dbus find them at runtime?
If dbus does not know where to find them, then they are useless.
Maybe some environment variable needs to be set?

Reply:
D-Bus can read values from `CMAKE_INSTALL_<dir>`, but `cmake.install()` installs all definitions in the default path (`<package_folder>/share`, `<package_folder>/etc` ...) instead of `<package_folder>/res/share`, etc. I tried to specify definitions in my _configure_cmake method, in build and package (before `cmake.install()` ) methods separately, but seems I don't know how it work exactly.
Also, I notice that in conan documentation (https://docs.conan.io/en/latest/reference/build_helpers/cmake.html?highlight=cmake_install_datarootdir) datarootdir sets to the default folder, but in conan hooks datarootdir, sysconfdir, localstatedir with default paths is wrong. This is very strange behaviour in my opinion, why not just set this in the res folder by default in the conan itself?
So, I can specify GNU install dirs in my _configure_cmake like this:
```
self._cmake.definitions["CMAKE_INSTALL_SYSCONFDIR"] =  os.path.join(self.package_folder, "res", "etc")
self._cmake.definitions["CMAKE_INSTALL_LOCALSTATEDIR"] =  os.path.join(self.package_folder, "res", "var")
self._cmake.definitions["CMAKE_INSTALL_DATAROOTDIR"] =  os.path.join(self.package_folder, "res", "share")
```
And dbus understand all of this and print:
```
install datadir: /home/semyon/.conan/data/dbus/1.12.20/_/_/package/d10203385f1b609816fdb837a88c3e1c4e81f9de1c4e81f9f/res/share
```
...
But anyway instalation sets definition paths to the default (`-- Installing: /home/semyon/.conan/data/dbus/1.12.20/_/_/package/d10203385f1b609816fdb837a88c3de1c4e81f9f/share/dbus-1/system.d`), therefore anything installs in `pakcage_folder` and I should move them in `res`. I'll be very glad if you can help me with this situation and descripe what I'm doing wrong?

Reply:
I checked the cmake script of dbus and it looks like the `share` folder is hardcoded.
So you're using the correct approach.